openid-4-verifiable-presentations-1_0.md

```diff
@@ -542,6 +542,266 @@ The Wallet then validates the request as specified in OAuth 2.0 [@RFC6749].
 
 If the Verifier responds with any HTTP error response, the Wallet MUST terminate the process.
 
+# Verifiable Presentation Query Language {#vp_query}
+
+The Verifiable Presentation Query Language (VP Query) is a JSON-encoded query
+language that allows the Verifier to request the Wallet to present Verifiable
+Credentials that match the query. The query consists of a set of constraints
+that the requested Verifiable Credentials must satisfy. The Wallet evaluates the
+query against the Verifiable Credentials it holds and returns the Verifiable
+Presentations that match the query.
+
+A valid VP Query is defined as a JSON-encoded object with the following
+properties:
+
+`credentials`:
+: REQUIRED. An object. Each key in the object represents one
+Credential Query as defined in (#credential_query). The key is a string that
+uniquely identifies the Credential Query within the VP Query.
+
+`rules`:
+: OPTIONAL. An object defining rules that specify which combinations of
+credentials MAY be returned by the Wallet, as defined in (#credential_rules). If
+omitted, all credentials listed in `credentials` are requested for presentation.
+
+## Credential Query {#credential_query}
+
+A Credential Query is an object representing a request for one specific
+credential. It contains the following properties:
+
+`format`:
+: REQUIRED. A string that specifies the format of the requested
+Verifiable Credential. Valid Credential format identifier values are defined in
+Appendix A of [@!OpenID.VCI]. The value of this property MUST be one of the
+supported Credential formats as defined in the Wallet's metadata.
+
+Format-specific properties (using the format value as a key):
+: OPTIONAL.
+Additional properties requested by the Verifier that are specific to the
+Credential format, as defined in (#format_specific_properties). If omitted,
+no format-specific restrictions are placed on the requested Credential.
+
+`purpose`:
+: OPTIONAL. A purpose description array as defined in
+(#purpose_description).
+
+`claims`:
+: OPTIONAL. An array that specifies the claims that the requested
+Verifiable Credential must contain. Each entry in the array corresponds to at
+least one claim that MUST be returned by the Wallet to fulfill the credential
+request (unless specified otherwise by extensions to this specification). Each
+entry MUST be one of the following objects:
+
+- Claim query: An object describing a single claim that is to be returned, as
+  specified in (#claim_query).
+- An optionality rule: An object containing the following properties:
+  - `required` (REQUIRED): defines a number of claims that must be returned from
+    the `from` set of claims, as specified next.
+  - `from` (REQUIRED): An array; each entry is either a claim query or another
```

Comment:
I'm not sure I fully understand the proposal - could you please give an example?

Reply:
Hereby some examples on what it could look like. Option 1 is the current proposal, option 2 and 3 are alternatives. 

```
; option (1) What we have in the current proposal

claims = [+claim]   ; you can ask for multiple "claim", each "claim" MUST be satisfied

claim = claim_query / optionality_rule  ; claim_query OR optionality_rule
optionality_rule = {
    "required": uint,   ; number of claims to be returned from the 'from' set
    "from": claim       ; recursively contains the claim structure
}

claim_query = claim_query_json / claim_query_mdoc   ; claim_query_json OR claim_query_mdoc

claim_query_json = {
    "path": tstr
}

claim_query_mdoc = {
    "namespace": tstr,
    "claim_name": tstr

}


; option (2) For each claim, define the main element, then define alternates for that claim

claims = [+claim]   ; you can ask for multiple "claim", each "claim" MUST be satisfied

claim = {
    "claim_query" = claim_query,            ; claim_query as defined in (1)
    "alternate_claims" = [+ claim_query]    ; Alternate claims, in order of preference
}


; option (3) For each claim, define an array 

claims = [+claim]       ; you can ask for multiple "claim", each "claim" MUST be satisfied
claim = [+claim_query]  ; to satisfy the "claim" array, one element from the array (e.g. one claim_query) must be returned

```

Note, what is the best solution also depends on the outcome of the other discussions. Specifically the ones regarding intent_to_retain, if_present and whether to have the claim_query be polymorphic or to have different versions of the "claims" element, e.g. claims_json and claims_mdoc.

Reply:
As discussed yesterday, hereby some examples for these structures. Note that other solutions besides option 2 or option 3 may exist that are more appropriate, especially when we consider the other comments on this PR.

```
Example 3: given_name, age_over_18 OR birth_date OR age_birth_year

option(1)

claims = [
    { "path": ["given_name"] },
    {
        "required": 1,
        "from": [
            { "path": ["age_over_18"] },
            { "path": ["birth_date"] },
            { "path": ["age_birth_year"] }
        ] 
    }
]


option(2)

claims = [ 
    {
        "claim_query": { "path": ["given_name"] }   
    },
    {
        "claim_query": { "path": ["age_over_18"] },
        "alternate_claims": [
            { "path": ["birth_date"] },
            { "path": ["age_birth_year"] }
        ]  
    },  
]

option(3)

claims = [ 
    [ { "path": ["given_name"] } ]
    [             
        { "path": ["age_over_18"] },
        { "path": ["birth_date"] },
        { "path": ["age_birth_year"] }
    ]
]



Example 4: given_name, zip_code OR (city AND state)

option(1)

claims = [
    { "path": ["given_name"] },
    {
        "required": 2,
        "from": [
            { "path": ["zip_code"] },
            {
                "required": 1,
                "from": [
                    { "path": ["city"] },
                    { "path": ["state"] },
                ]
            }
        ] 
    }
]

option(2)

claims = [ 
    {
        "claim_query": { "path": ["given_name"] }   
    },
    {
        "claim_query": { "path": ["zip_code"] },
        "alternate_claims": [
            { "path": ["city"] },
        ]  
    },  
    {
        "claim_query": { "path": ["zip_code"] },
        "alternate_claims": [
            { "path": ["state"] },
        ]  
    }
]

option(3)

claims = [ 
    [ { "path": ["given_name"] } ]
    [             
        { "path": ["zip_code"] },
        { "path": ["city"] },
    ],
    [             
        { "path": ["zip_code"] },
        { "path": ["state"] },
    ]
]
```

Reply:
Does `,` in the above mean AND/required? I think the description text should be more specific and informative.
Also, does the `Example 4` mean this `given_name AND (zip_code OR (city AND state))`
seems like in option 3 `zip_code OR city` and `zip_code OR state` could result in zip_code and zip_code. Resulting in not returning city or state.

Can we redefine option 3 by swapping the meaning of things in the 1st and 2nd level `[ ]`?
E.g. given_name AND (zip_code OR (city AND state))
```
redefining option 3
claims = [ 
    [ 
        { "path": ["given_name"] },
       { "path": ["zip_code"] },
    ]
    [             
       { "path": ["given_name"] },
       { "path": ["state"] },
       { "path": ["city"] },
    ],
]
```

This would make it simpler as entries in the root level `[ ]` would mean OR and the 2nd level will mean MUST.
In the above it translates to `(given_name and zipcode) OR (given_name and zip_code and city)`

Similarly for `Example 3: given_name and (age_over_18 OR birth_date OR age_birth_year)`
```
redefining option 3
claims = [
  [ 
        { "path": ["given_name"] },
       { "path": ["age_over_18"] },
  ],
 [ 
        { "path": ["given_name"] },
       { "path": ["birth_date"] },
  ],
  [ 
        { "path": ["given_name"] },
       { "path": ["age_birth_year"] },
  ],
```
This also corresponds to all the potential credentials User can consent to send. So it won't result in cases where `zip_code and zip_code` are sent.
I recognize that this could cause the query to be longer than other methods in few cases.

Reply:
I think it would be more clear to use a notation (req) instead of `and` 
as in boolean "zip_code" and "zip_code" translates to just 1 "zip_code"

https://grace.bluegrass.kctcs.edu/~kdunn0001/files/Simplification/4_Simplification_print.html

Where as in (req)zip_code (req)zip_code it is sent twice?